README.rst
pluck: Quickly pluck "fields" from a list of values
===================================================

.. image:: https://travis-ci.org/nvie/pluck.png
   :target: https://travis-ci.org/nvie/pluck


pluck is the simplest way of plucking "fields" from an iterable of values.
"Fields" are either ``item.field`` or ``item[field]``.  Pluck tries both,
in that order.  If nothing is found, and no default value is specified, it
throws an exception.


Usage
=====

The package consists of one module consisting of two functions::

   from pluck import pluck, ipluck

``ipluck`` is just the iterable version of ``pluck``.  Use it like this::

   pluck(iterable, key)

or::

   pluck(iterable, *keys)


Examples
========

A simple example first.  Say you have a list of datetimes::

   >>> from pluck import pluck
   >>> dates = [
   ...     datetime(2012, 10, 22, 12, 00),
   ...     datetime(2012, 10, 22, 15, 14),
   ...     datetime(2012, 10, 22, 21, 44),
   ... ]
   >>> pluck(dates, 'day')
   [22, 22, 22]
   >>> pluck(dates, 'hour')
   [12, 15, 21]

It also works on dictionary-like access (``__getitem__``)::

    >>> objects = [
    ...      {'id': 282, 'name': 'Alice', 'age': 30, 'sex': 'female'},
    ...      {'id': 217, 'name': 'Bob', 'age': 56},
    ...      {'id': 328, 'name': 'Charlie', 'age': 56, 'sex': 'male'},
    ... ]
    >>> pluck(objects, 'name')
    ['Alice', 'Bob', 'Charlie']
    >>> pluck(objects, 'age')
    [30, 56, 56]

You can also combine these into a single pluck::

   >>> pluck(objects, 'name', 'age')
   [('Alice', 30), ('Bob', 56), ('Charlie', 56)]


Defaults
========

You can specify default values, too.  By default, ``pluck`` will throw an
exception when a "field" does not exist::

   >>> pluck(objects, 'sex')
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
     File "pluck.py", line 104, in pluck
         return list(ipluck(iterable, *keys, **kwargs))
     File "pluck.py", line 49, in getter
         raise ValueError('Item %r has no attr or key for %r' % (item, key))
     ValueError: Item {'age': 56, 'id': 217, 'name': 'Bob'} has no attr or key for 'sex'

To instead fill these places with a default value, use this::

   >>> pluck(objects, 'sex', default='unknown')
   ['female', 'unknown', 'male']

When you specify multiple keys, you need to use the ``defaults`` (plural!)
keyword argument instead::

   >>> pluck(objects, 'name', 'sex', defaults={'sex': 'unknown'})
   [('Alice', 'female'), ('Bob', 'unknown'), ('Charlie', 'male')]


Iterator, rather?
=================

Use ``ipluck`` if you'd rather wanna have an iterator::

    >>> from pluck import ipluck
    >>> ipluck(large_stream_of_items, 'name')
    <itertools.imap object at 0x10c7515d0>

``pluck`` is equivalent to ``list(ipluck(...))``.


HISTORY.rst
History
=======

0.2 (2012-10-22)
------------------
   - Add Python 3 compatibility
   - Improve documentation

0.1 (2012-10-22)
------------------
   - Initial release.


test_pluck.py
import unittest
from datetime import datetime
from pluck import pluck


class TestPluck(unittest.TestCase):
    def test_empty_iterables(self):
        assert pluck([], 'foo') == []
        assert set(pluck(set(), 'foo')) == set()

    def test_getattr_based_plucks(self):
        l = [
                datetime(2012, 10, 1, 3),
                datetime(2012, 10, 15, 4),
                datetime(2012, 10, 30, 8),
            ]
        assert pluck(l, 'year') == [2012, 2012, 2012]
        assert pluck(l, 'day') == [1, 15, 30]
        assert pluck(l, 'hour') == [3, 4, 8]

    def test_getitem_based_plucks(self):
        l = [
                {'id': 282, 'name': 'Alice', 'age': 30},
                {'id': 217, 'name': 'Bob', 'age': 56},
                {'id': 328, 'name': 'Charlie', 'age': 56},
                {'id': '167'},
            ]
        assert pluck(l, 'id') == [282, 217, 328, '167']
        self.assertRaises(ValueError, pluck, l, 'name')
        assert pluck(l, 'name', default=None) == ['Alice', 'Bob', 'Charlie', None]
        assert pluck(l, 'name', default='Mr. X') == ['Alice', 'Bob', 'Charlie', 'Mr. X']
        assert pluck(l, 'age', default=None) == [30, 56, 56, None]

    def test_get_multiple_fields(self):
        l = [
                {'id': 282, 'name': 'Alice', 'age': 30},
                {'id': 217, 'name': 'Bob', 'age': 56},
                {'id': 328, 'name': 'Charlie', 'age': 56},
                {'id': '167'},
            ]
        self.assertRaises(ValueError, pluck, l, 'id', 'name')
        assert pluck(l, 'id', 'name', defaults={'name': 'Mr. X'}) == \
               [(282, 'Alice'), (217, 'Bob'), (328, 'Charlie'), ('167', 'Mr. X')]

        # Specifying a default works, as long as all other fields are found
        self.assertRaises(ValueError, pluck, l, 'name', 'age')

        # This still fails, because there are missing age fields
        self.assertRaises(ValueError, pluck, l, 'name', 'age', defaults={'name': 'Mr. X'})

        assert pluck(l, 'id', 'age', defaults={'age': 100}) == \
               [(282, 30), (217, 56), (328, 56), ('167', 100)]

    def test_works_with_iterables(self):
        l = iter([
                {'id': 282, 'name': 'Alice', 'age': 30},
                {'id': 217, 'name': 'Bob', 'age': 56},
                {'id': 328, 'name': 'Charlie', 'age': 56},
                {'id': 167},
            ])
        assert pluck(l, 'id', 'name', defaults={'name': 'Mr. X'}) == \
               [(282, 'Alice'), (217, 'Bob'), (328, 'Charlie'), (167, 'Mr. X')]


if __name__ == '__main__':
    unittest.main()


pluck.py
## -*- coding: utf-8 -*-
"""
pluck
=====

pluck is the simplest way of plucking "fields" from an iterable of values.
"Fields" are either ``item.field`` or ``item[field]``.  Pluck tries both,
in that order.  If nothing is found, and no default value is specified, it
throws an exception.
"""
import sys

PY3 = sys.version_info >= (3, 0)
if not PY3:
    from future_builtins import map, zip

from itertools import tee
import operator

__title__ = 'pluck'
__version__ = '0.2'
__author__ = 'Vincent Driessen'
__license__ = 'BSD'
__copyright__ = 'Copyright 2012 Vincent Driessen'


class _Const(object):
    pass

FAIL = _Const()


def pluck_single(iterable, key, default=FAIL):
    return list(ipluck_single(iterable, key, default=default))


def ipluck_single(iterable, key, default=FAIL):
    attrgetter = operator.attrgetter(key)
    itemgetter = operator.itemgetter(key)

    def getter(item):
        try:
            return attrgetter(item)
        except AttributeError:
            pass

        try:
            return itemgetter(item)
        except KeyError:
            pass

        if default is not FAIL:
            return default

        raise ValueError('Item %r has no attr or key for %r' % (item, key))

    return map(getter, iterable)


def ipluck_multiple(iterable, defaults, *keys):
    if len(keys) > 1:
        iters = tee(iterable, len(keys))
    else:
        iters = (iterable,)
    iters = [ipluck_single(it, key, default=defaults.get(key, FAIL)) for it, key in zip(iters, keys)]
    return zip(*iters)


def ipluck(iterable, key, *keys, **kwargs):
    """Iterator version of pluck()."""
    if len(keys) > 0:
        defaults = kwargs.pop('defaults', {})
        return ipluck_multiple(iterable, defaults, key, *keys)
    else:
        default = kwargs.pop('default', FAIL)
        return ipluck_single(iterable, key, default=default)


def pluck(iterable, *keys, **kwargs):
    """pluck is the simplest way of plucking "fields" from an iterable of
    values.  "Fields" are either ``item.field`` or ``item[field]``.  Pluck
    tries both, in that order.  If nothing is found, and no default value is
    specified, it throws an exception.

    Examples:

        >>> dates = [
        ...     datetime(2012, 10, 22, 12, 00),
        ...     datetime(2012, 10, 22, 15, 14),
        ...     datetime(2012, 10, 22, 21, 44),
        ... ]
        >>> objects = [
        ...      {'id': 282, 'name': 'Alice', 'age': 30, 'sex': 'female'},
        ...      {'id': 217, 'name': 'Bob', 'age': 56},
        ...      {'id': 328, 'name': 'Charlie', 'age': 56, 'sex': 'male'},
        ... ]

        >>> pluck(dates, 'hour')
        [12, 15, 21]
        >>> pluck(objects, 'name')
        ['Alice', 'Bob', 'Charlie']
        >>> pluck(objects, 'name', 'age')
        [('Alice', 30), ('Bob', 56), ('Charlie': 56)]
        >>> pluck(objects, 'name', 'sex')
        ValueError: item {'id': 217, 'name': 'Bob', 'age': 56} has no attr or key 'sex'
        >>> pluck(dates, 'name', 'sex', defaults={'sex': 'unknown'})
        [('Alice', 'female'), ('Bob', 'unknown'), ('Charlie': 'male')]

    """
    return list(ipluck(iterable, *keys, **kwargs))


__all__ = ['pluck', 'ipluck']
